Execute user scripts in a subshell

TomekTrzeciak opened the pull request:
<!-- Complete this Pull Request template. -->

<!-- Significant PRs should address an existing Issue. Choose one: -->

These changes close #2043

<!-- The following requirements must be satisfied (with "[x]"). -->
<!-- Mark the PR as a Draft if all requirements are not yet satisfied. -->

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
<!-- choose one: -->
- [x] Appropriate tests are included (unit and/or functional).
- [ ] Already covered by existing tests.
- [ ] Does not need tests (why?).
<!-- choose one: -->
- [ ] Appropriate change log entry included.
- [ ] No change log entry required (why? e.g. invisible to users).
<!-- choose one: -->
- [x] (master branch) I have opened a documentation PR at cylc/cylc-doc#108.
- [ ] (7.8.x branch) I have updated the documentation in this PR branch.
- [ ] No documentation update required.

TomekTrzeciak commented:
List of failing tests from travis:
- [x] `tests/cylc-poll/10-basic-compat.t` - passes for me locally
- [ ] `tests/cylc-poll/11-event-time.t` - relies on in-process execution
- [ ] `tests/cylc-poll/15-job-st-file-no-batch.t` - relies on in-process execution
- [ ] `tests/hold-release/12-hold-then-retry.t` - relies on in-process execution
- [ ] `tests/job-file-trap/00-sigusr1.t` - I don't understand what this one is testing and how
- [ ] `tests/job-file-trap/02-pipefail.t` - easy fix, `CYLC_JOB_EXIT` is `ERR` instead of `EXIT`
- [ ] `tests/jobscript/12-err-script.t` - relies on in-process execution
- [ ] `tests/jobscript/19-exit-script.t` - relies on in-process execution
- [x] `tests/reload/17-graphing-change.t` - passes for me locally
- [ ] `tests/restart/00-pre-initial.t` - relies on in-process execution (`wait "${CYLC_TASK_MESSAGE_STARTED_PID}"` in task script, easy fix)
- [ ] `tests/shutdown/09-now2.t` - relies on in-process execution
- [ ] `tests/shutdown/14-no-dir-check.t` - relies on in-process execution

Tests that rely on in-process execution either muck about with traps from within task script or make use of arguments passed to the trap script. This tests internal details of implementation rather than functionality and it would be best to redesign these. I would need help with this, though, as I don't even understand the purpose of some of them, so please help.

Interestingly, the one test that I would expect to fail (`tests/events/34-task-abort.t` calling `cylc__job_aboort` shell function) actually succeeds 😕

TomekTrzeciak commented:
The problem from issue #2043 with signal trap not triggering until child process termination is not solved here, but based on https://unix.stackexchange.com/a/146770 changing `( user_script )` to `( user_script ) & wait $!` should allow the trap function to trigger immediately.

Additionally, adding `kill -s ${signal} 0` in the trap function should propagate the signal to the current process group and kill the user script child process (unless the child sets its own trap for this signal).

kinow commented on cylc/flow/etc/job.sh:
```diff
@@ -121,11 +119,16 @@ cylc__job__main() {
-    done
+    # Env-Script, User Environment, Pre-Script, Script and Post-Script
+    # Execute in the subshell to protect job script environment from
+    # interferance and put in the background to allow signal traps to trigger.
```
s/interferance/interference

kinow commented on cylc/flow/etc/job.sh:
```diff
@@ -121,11 +119,16 @@ cylc__job__main() {
+            cylc__job__run_inst_func "${func_name}"
+        done
+    ) &
+    wait $!
```
@TomekTrzeciak I am probably the one with less Shell knowledge in the team, but had some spare time and decided to have a look - as I know others are busy at the moment too.

Checked out your branch, then chose one of the failing tests for testing: `tests/cylc-poll/11-event-time.t`. It fails to run the suite reference check.

To identify what could have caused it, and as your PR is changing only a few lines of a single file, I went "toggling" your changes (i.e. reverting parts of it).

This line seems to cause the issue for this test on my environment. From what I understand, we start a subshell with the `(` and `)`. Send it to the background with `&`, then wait until this process is done with `wait $!`.

It looks good to me. But the test passes only when waiting for all the processes with `wait` (without the PID) instead of `wait $`. Not really sure why? :confused:

TomekTrzeciak replied:
Many thanks for looking into this @kinow. These tests will not work without modification as they rely on modifying/triggering traps in ways that are not accessible from a subshell/background process. However, I'm almost done with fixing all the failing tests. The key idea is to use `init-script` for setting up the trap related stuff, since this script is executed in the same shell as the job script.

kinow replied:
Will re-run the tests once you've finished working on this branch then @TomekTrzeciak. At least I am learning a bit more of Shell reading your changes.

TomekTrzeciak commented on tests/jobscript/12-err-script/suite.rc:
```diff
@@ -8,9 +8,8 @@
 [runtime]
     [[foo]]
+        init-script = FOO="foo bar baz qux"
         err-script = echo "$1 ${FOO}"
```
With this PR it is no longer possible to access user environment or script variables from `err-script` and `exit-script` (by design) and instead `init-script` has to be used for that.

TomekTrzeciak commented:
OK, I think this is ready for review. There is one failed test, but it passes for me locally and it did pass on travis on previous run too.

TomekTrzeciak commented:
Another test failing that hasn't failed before (`tests/pre-initial/07-simple-messaging.t`). It works for me locally and it's a relatively simple test that only runs reftest of the following suite:

```
[cylc]
   UTC mode = True
[scheduling]
    initial cycle point = 20100808T00
    final cycle point = 20100809T00
    [[graph]]
        T00 = "foo[-P1D]:restart1 => foo"
[runtime]
    [[foo]]
        script = """
wait
cylc message "restart files ready"
"""
        [[[outputs]]]
            restart1 = "restart files ready"
```

The difference to master is that the `wait` command, which makes sure the background `cylc message started` completes, has no effect here and the custom message can arrive before it. I tested what happens then and the user message is just ignored (based on suite log) and doesn't trigger the next task. This creates a race condition for user code that relies on custom message triggers. I recon that it might be necessary to use trap/wait tricks to set up the right conditions for some tests, but I guess normal user code should not require them (CC: @hjoliver).

Are failed tests rerun on travis? I remember they used to be.

kinow commented:
>Are failed tests rerun on travis? I remember they used to be.

They are, but sometimes these tests fail again. I've kicked Travis just in case.

kinow commented:
The job that failed is now passing. I'll triggering the flaky tests job now (should take another 40ish minutes).

kinow commented:
@TomekTrzeciak Travis build passed. Great job ! 👏

TomekTrzeciak commented:
> @TomekTrzeciak Travis build passed, just needed a few kicks. Great job ! 👏

@kinow, many thanks again! I was hoping to see the flaky tests will run fine and I'm glad they did. But the fact that the tests pass doesn't mean that there are no race conditions there. I believe the one I've described above is a genuine edge case that also affects user code. While I'm not able to check all the tests, I've gone with with a fine comb (err, `grep` that is) over all uses of trap and wait in the tests and adapted them to behave as close to the original code as possible.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -140,6 +135,34 @@ cylc__job__main() {
+    done &
+    CYLC_TASK_SCRIPT_PID=$!
+    wait "${CYLC_TASK_SCRIPT_PID}"
+}
```
This essentially implements the suggested solution from #2043. Using background subshell and `wait` allows the trap to execute immediately after receiving the signal rather than after the execution of current command completes.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -193,6 +216,14 @@ cylc__job_finish_err() {
+        kill -s "${signal}" -- "-$$" 2>'/dev/null' ||
+        kill -s "${signal}" "${CYLC_TASK_SCRIPT_PID}" 2>'/dev/null' || true
+        wait  # in case child user script traps the signal
+    fi
```
The logic here is to try to resend the signal to the entire process group (will succeed if jobscript shell is a group leader - should this be always the case?) or, if that fails, to the backgrounded user script subshell. User script is free set its own traps, including setting them to ignored for itself and/or its children. Then we wait for user script process to finish, so that there is no risk of running it concurrently with `err-script` (I think this is a sensible thing to do).

TomekTrzeciak replied:
Actually, there is problem with this. When bash receives a signal during `wait` command it returns from that command with exit code 128 + signal number. Because the error code is non-zero, ERR trap triggers before the signal trap and, due to `exit 1` in there, the signal trap never gets executed. This is pretty annoying, because distinguishing between signal and error exit becomes more complicated.

TomekTrzeciak replied:
The above is still not quite correct. It is what one might expect (at least what I would expect based on bash docs), but reality is more complicated. Just have a look at this simple test:
```
$ bash -c 'for sig in ERR EXIT TERM; do trap "echo $sig:\$?" $sig; done; f(){ sleep 5 & wait $!; }; f' & sleep 1; kill -- $!; wait
TERM:143
ERR:143
EXIT:143
$ bash -ec 'for sig in ERR EXIT TERM; do trap "echo $sig:\$?" $sig; done; f(){ sleep 5 & wait $!; }; f' & sleep 1; kill -- $!; wait
TERM:143
EXIT:143
$ bash -ec 'for sig in ERR EXIT TERM; do trap "echo $sig:\$?" $sig; done; sleep 5 & wait $!' & sleep 1; kill -- $!; wait
ERR:143
TERM:143
EXIT:143
$ bash --version
GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)
```
So the order of traps depends on whether or not the function is used. Moreover, doing the same with bash 3.2 I get TERM trap executed first in all these cases, so this behaviour is not even consistent bash between versions (at least the ones I have access to test with).

The conclusion from that is that one cannot trust the order of trap execution in bash. Special care is therefore needed to make sure that the right traps are triggered and in the right order. I think I have a working solution to this now and will push it up soon for discussion.

And now for the real kicker. While going down this rabbit hole I've come across this lovely post:
https://lists.gnu.org/archive/html/bug-bash/2012-12/msg00093.html

The gist of it is to never, ever use functions in conditionals in combination with `set -e` (unless you specifically design them for this) or you might be in for a really bad surprise (spoiler alert: POSIX mandates that `set -e` is simply disabled for anything executing within the conditional context with disastrous consequences for functions and compound commands). But read the linked post, it's worth it.

kinow replied:
I think if I had to choose between C, Fortran, and Shell, Shell would probably be my last option :eyes: Great detective work @TomekTrzeciak :clap:

TomekTrzeciak replied:
More of bash woes:
```
$ bash -ec 'f() { trap "echo xxx" EXIT; false; };  f & wait $!'; echo $?
xxx
1
$ bash -c 'f() { trap "echo xxx" EXIT; false; };  f & wait $!'; echo $?
1
$ bash -c 'f() { trap "echo xxx" EXIT; false; exit; };  f & wait $!'; echo $?
xxx
1
```
Looks like bash remembers to execute EXIT traps only some of the time. I get this with bash 4.2 and 3.2 alike.

TomekTrzeciak replied:
Even more bash insanity:
```
$ bash -c 'f() { for S in ERR EXIT; do trap "echo ${S}_CODE=\$?" ${S};  done; (exit 42); };  (f) & wait $!'
ERR_CODE=42
EXIT_CODE=42
$ bash -ec 'f() { for S in ERR EXIT; do trap "echo ${S}_CODE=\$?" ${S};  done; (exit 42); };  (f) & wait $!'
ERR_CODE=42
EXIT_CODE=1
$ bash --version
GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)
```
Somewhere between handling ERR and EXIT traps `bash -e` decides to change the status code! WTF?

(This works as expected on bash 3.2 or if you replace `(f) &` with `f &`, expect that the error trap is not executed then unless you also have `set -e`.)

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -193,6 +216,14 @@ cylc__job_finish_err() {
         wait "${CYLC_TASK_MESSAGE_STARTED_PID}" 2>'/dev/null' || true
     fi
     cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" || true
```
This could possibly be run as a background process similarly to CYLC_TASK_MESSAGE_STARTED_PID to let the rest of the trap code carry on without waiting on network communication.

TomekTrzeciak commented:
@matthewrmshin, you might be one of the most most knowledgeable people about this issue, would be good if you could chime in here.

hjoliver commented:
@TomekTrzeciak - I'm just back from an overseas trip.  Good to see all the action on this, thanks.  To help short-cut the catch-up process for me, could you please give a quick summary of where you're at now on this PR branch (what issues if any remain to be solved, etc.)? (I see you've pushed a couple more commits since the last comments).

matthewrmshin commented:
Note: I've booked some time with @TomekTrzeciak on Monday to have a chat about this.

TomekTrzeciak commented:
__Executive summary for @hjoliver 😉__

This PR  implements the suggested solution from #2043. User scripts are executed via background subshell, which is `wait`ed for (otherwise shell will not execute the trap code until the foreground commaind finishes).

I've discovered inconsistencies and bugs in how different versions bash handle traps and `set -e` error conditions. I've spent a lot of time to work around them and deliver consistent behaviour on the two versions I have access to (bash 4.2 and 3.2). If anyone else could run the tests on other versions, that would be good. The short list of relevant tests, that either needed to be adapted for or proved to be sensitive to this change at some point along the way, is included below. (The two failing tests in the last round - `tests/reload/17-graphing-change.t` and `tests/cylc-scan/02-sigstop.t` - are due to travis playing up. Could someone give it a kick?)

I provide more detail in inline comments, but the main features/caveats are:
- User scripts (`env-script`, user environment, `pre-script`, `script` and `post-script`) are isolated in a separate process, so any environment changes, traps, etc. are not visible from the main jobscript and are safe to do now.
- Traps should be triggered in a predictable and consistent order (signal trap first, if signal is received by the main script, `ERR` trap on child error and `EXIT` on child abnormal termination (exit code > 128) due to some other signal like SIGPIPE in command pipeline).
- Trap handler will resend the signal to a process group (if it's a leader) or child user script (if no a leader - is this ever a case?). Might be worth considering to re-enable `TERM` trap on slurm (see #1287) if this change is accepted.
- The exit code of user script process is preserved (including signal exits) and made available to the trap code. This isn't used ATM, but could be logged or send back to cylc.
- `exit-script` and `err-script` will no longer share environment with user scripts and any common variables have to be set in `init-script`.
- `cylc__job_abort` does appear to work, but cannot short-circuit execution any more and fail message will be sent afterwords (which is ignored by cylc). Not sure if this is a problem, but I've noticed that if aborted message is not delivered, task poll is not capable to work out that the abort happend and treats it as a generic fail. This happens on master too, so appear to be a limitation/bug in current implementation that is unrelated to this change.

__List of relevant tests__

- [x] `flakytests/cylc-kill/02-submitted.t`
- [x] `flakytests/cylc-poll/03-poll-all.t`
- [x] `flakytests/cylc-poll/16-execution-time-limit.t`
- [x] `flakytests/cylc-show/04-multi.t`
- [x] `flakytests/cylc-take-checkpoints/00-basic.t`
- [x] `flakytests/events/01-task.t`
- [x] `flakytests/events/39-task-event-template-all.t`
- [x] `flakytests/execution-time-limit/00-background.t`
- [x] `flakytests/execution-time-limit/01-at.t`
- [x] `flakytests/execution-time-limit/04-poll.t`
- [x] `flakytests/job-submission/19-chatty.t`
- [x] `flakytests/restart/19-checkpoint.t`
- [x] `flakytests/shutdown/02-no-dir.t`
- [x] `tests/cylc-message/02-multi.t`
- [x] `tests/cylc-poll/02-task-submit-failed.t`
- [x] `tests/cylc-poll/10-basic-compat.t`
- [x] `tests/cylc-poll/11-event-time.t`
- [x] `tests/cylc-poll/15-job-st-file-no-batch.t`
- [x] `tests/database/04-lock-recover.t`
- [x] `tests/events/28-inactivity.t`
- [x] `tests/events/34-task-abort.t`
- [x] `tests/hold-release/12-hold-then-retry.t`
- [x] `tests/hold-release/19-no-reset-prereq-on-waiting.t`
- [x] `tests/job-file-trap/00-sigusr1.t`
- [x] `tests/job-file-trap/02-pipefail.t`
- [x] `tests/job-submission/08-activity-log-host.t`
- [x] `tests/jobscript/12-err-script.t`
- [x] `tests/jobscript/19-exit-script.t`
- [x] `tests/jobscript/19-exit-script.t`
- [x] `tests/reload/19-remote-kill.t`
- [x] `tests/reload/20-stop-point.t`
- [x] `tests/restart/00-pre-initial.t`
- [x] `tests/restart/28-execution-timeout.t`
- [x] `tests/shutdown/08-now1.t`
- [x] `tests/shutdown/09-now2.t`
- [x] `tests/shutdown/12-bad-port-file-check.t`
- [x] `tests/shutdown/13-no-port-file-check.t`
- [x] `tests/shutdown/14-no-dir-check.t`

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -121,11 +119,23 @@ cylc__job__main() {
+            (exit "$ret_code")
+            return
+        fi
+    }
```
This is a pretty crafty solution but it's necessary to use a backgrounded function rather than subshell to preserve the error code on bash 4.2.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -140,6 +150,46 @@ cylc__job__main() {
+# Disable selected or all (if no arguments given) fail traps.
+# Globals:
+#   CYLC_FAIL_SIGNALS
+cylc__job__disable_fail_signals() {
```
I figured it's cleaner to add a function for this than directly muck around with CYLC_FAIL_SIGNALS variable in the tests.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -182,26 +233,41 @@ cylc__job__run_inst_func() {
+        kill -s "${signal}" -- "-$$" 2>'/dev/null' ||
+        kill -s "${signal}" -- "${CYLC_TASK_USER_SCRIPT_PID}" 2>'/dev/null' || true
+    fi
+    cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" &
```
Time is of essence after receiving a signal, so we I think it's sensible to overlap network communication with `err-script`.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -182,26 +233,41 @@ cylc__job__run_inst_func() {
     if "${run_err_script}"; then
+        if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]]; then
+            # User script might trap the signal, wait for it before err-script
+            wait  "${CYLC_TASK_USER_SCRIPT_PID:-}" 2>'/dev/null' || true
```
I assume we want this here to make sure `err-script` is not run concurrently with user scripts in case they trap or ignore the signal.

TomekTrzeciak commented on tests/events/16-task-event-job-logs-register-globalcfg/suite.rc:
```diff
@@ -1,21 +0,0 @@
-#!jinja2
-[meta]
-    title=Task Event Job Log Retrieve
```
The test using this suite has been removed in #1976 a long time ago.

TomekTrzeciak commented on flakytests/execution-time-limit/00-background.t:
```diff
@@ -34,7 +34,7 @@ suite_run_fail "${TEST_NAME_BASE}-run" \
 LOGD="$(cylc get-global-config --print-run-dir)/${SUITE_NAME}/log/job/1/foo"
 grep_ok '# Execution time limit: 5.0' "${LOGD}/01/job"
-grep_ok 'CYLC_JOB_EXIT=\(ERR\|XCPU\)' "${LOGD}/01/job.status"
+grep_ok 'CYLC_JOB_EXIT=XCPU' "${LOGD}/01/job.status"
```
This test is a manifestation of unreliability of trap execution order in bash. After this PR we should consistently see an exit via signal trap.

TomekTrzeciak commented on tests/job-file-trap/02-pipefail.t:
```diff
@@ -18,20 +18,26 @@
+CYLC_JOB_EXIT=EXIT
+__STATUS__
+grep_ok 'CYLC_JOB_EXIT_TIME=' "${T2_STATUS_FILE}"
+
```
Normal errors should now consistently trigger ERR trap, while abnormal termination (exit code > 128) should go via exit trap.

hjoliver commented:
@TomekTrzeciak - thanks for the excellent summary. It sounds good to me (and we can probably use it as documentation for how job scripts work, with a few minor tweaks).  I'll await the outcome of your discussion with @matthewrmshin before taking this further, but if Matt's happy (as most recent job script expert) we should be good to go.

matthewrmshin commented:
Met with @TomekTrzeciak today, who walked me through the sticking points and quirks. He also mentioned that he would be happy to help maintain this change should anything goes wrong. I cannot see any big issues with this change.

hjoliver commented:
@matthewrmshin - brilliant, and thanks for pitching in on this, much appreciated :+1: 

@TomekTrzeciak - to finish his up (with reference to the requirements check-list):
- Do you think you have enough test coverage already? (given that many existing tests cover job scripts anyway)
- Do you want to back-port this to 7.8.x (we will be making at least one more release, 7.8.5) or are you happy to wait for Cylc 8 to use this?  Not 100% sure, but probably just a trivial code transplant to back-port.
- Could you briefly adapt your executive summary above for the User Guide? (target audience: anyone who needs to know exactly how job scripts work). Current rather minimal docs here: https://cylc.github.io/doc/built-sphinx/task-implementation.html

(For Cylc 8+ docs are now at the cylc/cylc-doc repo; for 7.8.x they're still in cylc/cylc).

TomekTrzeciak commented:
> * Do you think you have enough test coverage already? (given that many existing tests cover job scripts anyway)

I think there is a decent amount of test coverage already and I've also modified and extended a few tests to make them even stricter.

> * Do you want to back-port this to 7.8.x (we will be making at least one more release, 7.8.5) or are you happy to wait for Cylc 8 to use this?  Not 100% sure, but probably just a trivial code transplant to back-port.

I think this change is more appropriate for a major release, because it introduces some (small that they may be, but still...) backward incompatibilities. (Mainly, `exit-script` and `err-script` is not sharing the environment with the main user scripts, which will break any suites relying on this, plus `cylc__job_abort` behaviour differences, which are fixable, but see below.)

> * Could you briefly adapt your executive summary above for the User Guide? (target audience: anyone who needs to know exactly how job scripts work). Current rather minimal docs here: https://cylc.github.io/doc/built-sphinx/task-implementation.html

OK, I'll open a PR on cylc-doc for it.

The outstanding question I have is about `cylc__job_abort` functionality. I can fix it to behave like it does now, see the diff below, but I wonder if it's worth it.

Firstly, `cylc__job_abort` addresses a very niche use case of passing a custom message to the `failed handler`. Users can instead simply use `cylc message "CRITICAL: ..." && exit 1` + `critical handler` (as suggested in the documentation), which achieves pretty much the same effect, so why bother.

Secondly, there is currently a problem with `cylc__job_abort` in that it doesn't work correctly with task poll (in case `cylc message ...` is blocked or fails). This is because the `aborted/` message prefix is stripped out and never written to `job.status` and so task poll can't work it out and treats the abort situation as a generic failure instead, in which case the custom message never gets passed to the `failed handler`. That would need to be fixed outside of this PR.

The fix below keeps things behaving as they do now, but is not elegant in that it intertwines details of `cylc message` and `job.status` with jobscript implementation. My instinct would be to just drop support for `cylc__job_abort` altogether, which would also make the problem with task poll go away. But let me know which way you would prefer to go with this.

```diff
diff --git a/cylc/flow/etc/job.sh b/cylc/flow/etc/job.sh
index f76a9a2c9..e0959f58b 100644
--- a/cylc/flow/etc/job.sh
+++ b/cylc/flow/etc/job.sh
@@ -221,6 +221,7 @@ cylc__job__run_inst_func() {
 # Send message (and possibly run err script) before job exit.
 # Globals:
 #   CYLC_FAIL_SIGNALS
+#   CYLC_TASK_LOG_ROOT
 #   CYLC_TASK_USER_SCRIPT_PID
 #   CYLC_TASK_USER_SCRIPT_EXITCODE
 #   CYLC_VACATION_SIGNALS
@@ -250,6 +251,7 @@ cylc__job_finish_err() {
         kill -s "${signal}" -- "-$$" 2>'/dev/null' ||
         kill -s "${signal}" -- "${CYLC_TASK_USER_SCRIPT_PID}" 2>'/dev/null' || true
     fi
+    grep -q "^CYLC_JOB_EXIT=" "${CYLC_TASK_LOG_ROOT}.status" ||
     cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" &
     if "${run_err_script}"; then
         if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]]; then
@@ -268,7 +270,7 @@ cylc__job_finish_err() {
 #   CYLC_TASK_USER_SCRIPT_EXITCODE
 cylc__job_abort() {
     CYLC_TASK_USER_SCRIPT_EXITCODE="1"
-    cylc__job_finish_err "EXIT" true "CRITICAL: aborted/\"${1}\""
+    cylc__job_finish_err "EXIT" false "CRITICAL: aborted/\"${1}\""
 }
 
 ###############################################################################
diff --git a/tests/events/34-task-abort.t b/tests/events/34-task-abort.t
index 690719679..ea6dbc887 100755
--- a/tests/events/34-task-abort.t
+++ b/tests/events/34-task-abort.t
@@ -1,7 +1,7 @@
 #!/bin/bash
 # THIS FILE IS PART OF THE CYLC SUITE ENGINE.
 # Copyright (C) 2008-2019 NIWA & British Crown (Met Office) & Contributors.
-# 
+#
 # This program is free software: you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation, either version 3 of the License, or
@@ -17,7 +17,7 @@
 #-------------------------------------------------------------------------------
 # Test job abort-with-message and interaction with failed handler.
 . "$(dirname "$0")/test_header"
-set_test_number 5
+set_test_number 6
 install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
 #-------------------------------------------------------------------------------
 run_ok "${TEST_NAME_BASE}-validate" cylc validate "${SUITE_NAME}"
@@ -41,5 +41,10 @@ grep_ok 'ONE' "${LOG}"
 # ...after abort
 grep_fail 'TWO' "${LOG}"
 #-------------------------------------------------------------------------------
+# Check only one CYLC_JOB_EXIT message has been written.
+JOB_STATUS="${SUITE_RUN_DIR}/log/job/1/foo/NN/job.status"
+run_ok "${TEST_NAME_BASE}-exit-count" \
+    test "$(grep -c '^CYLC_JOB_EXIT=' "$JOB_STATUS")" -eq '1'
+#-------------------------------------------------------------------------------
 purge_suite "${SUITE_NAME}"
 exit
diff --git a/tests/events/34-task-abort/suite.rc b/tests/events/34-task-abort/suite.rc
index 7d9374741..eaef25cb4 100644
--- a/tests/events/34-task-abort/suite.rc
+++ b/tests/events/34-task-abort/suite.rc
@@ -9,6 +9,7 @@
 [runtime]
     [[foo]]
         script = """
+cylc__job__wait_cylc_message_started
 echo ONE
 cylc__job_abort "ERROR: rust never sleeps"
 echo TWO"""
```

hjoliver commented:
@TomekTrzeciak - we discussed this in the most recent Cylc project meeting.  @dpmatthews flagged concerns at the critical nature of this change, along with the difficulty of testing all scenarios well before deploying it (remote jobs host, PBS etc.).  However, I think we all agreed this is a good change in principle, and it solves some outstanding problems, so it should go in for Cylc 8 - which leaves lots of time for shaking out problems ... if you can perhaps agree to be the job script expert going forward and follow-up on any problems that result?

I need to follow-up on `cylc__job_abort` - that went in for BOM use cases and is probably in use there now.  Beyond that, @oliver-sanders and I should do a closer review when time allows, then we should be able to merge this.

TomekTrzeciak commented:
@hjoliver, I'm happy to help maintain the script in case of problems. You know I do that for other pieces of work I've contributed anyway (e.g., EmPy).

Also happy for more people to review it. The more eyes on it the better as far as I'm concerned. That's why I did rope in @matthewrmshin here, as I know @dpmatthews is in a tough spot with resourcing at the moment. I might be able to test this change with bash 5.0 soon and @dpmatthews told me he might look more at testing with slurm, pbs and etc.

We have a collaborator from BOM on site (@tjtg), so I quizzed him about `cylc__job_abort` usage over there, but our quick search didn't turn up any answers, so I'll just wait on word from you about that. I don't mind this going either way, but just thought it's worth to flag this up for cylc 8 as it's a good opportunity to shed some craft.

TomekTrzeciak commented:
In the interest of moving this along, I've added support for `cylc__job_abort` in the last commit (can be easily taken off if needed). A small fix for bugginess of bash v3 was needed and I've added a new test for TERM trap in the user script (which wasn't possible before). I've also put up a doc PR.

I haven't managed to test with bash v5, the installation I had access to wasn't working properly. If anybody else has access to the latest bash and feels like checking this PR, it's enough to just run the list of tests from https://github.com/cylc/cylc-flow/pull/3440#issuecomment-559984970, but even without this I'm quite confident about this change by now.

Oh, and travis got stack again, could someone give it a kick.

hjoliver commented:
Thanks @TomekTrzeciak , sounds good. I'm just back from summer holidays here, will try to take a good look at this soon.

oliver-sanders commented:
This review appears to have stalled, we could do with making progress before the code becomes too stale.

hjoliver commented:
> This review appears to have stalled, we could do with making progress before the code becomes too stale.

Agreed, but we're pretty slammed with higher priorities right now.  Once delta-driven infinite tree is working we can relax a little and return to other important issues like this (soon!).

Fortunately this branch is only conflicting with a few tests at this point.

dpmatthews commented:
Presumably this PR should remove the special treatment of the TERM signal for Slurm. I don't think that's included yet.

TomekTrzeciak commented:
> Presumably this PR should remove the special treatment of the TERM signal for Slurm. I don't think that's included yet.

I didn't want to jump the gun on this before getting some sort of go ahead first.

TomekTrzeciak commented:
> > Presumably this PR should remove the special treatment of the TERM signal for Slurm. I don't think that's included yet.
> 
> I didn't want to jump the gun on this before getting some sort of go ahead first.

OK, I re-enabled TERM for Slurm and rebased to remove conflicts. All tests passed.

(How does one run tests locally now that cylc commands have moved to entry points? In the past I just added repo `bin` to `$PATH`.)

oliver-sanders commented:
> (How does one run tests locally now that cylc commands have moved to entry points? In the past I just added repo bin to $PATH.)

Conda

kinow commented:
Or if you use pip, just (re-)create a virtual environment and run `pip install -e .[all]`.

TomekTrzeciak commented:
I've just noticed a bunch of new conflicts here as a result of #3653. Would it make sense to get this in before #3654 (looks like a massive shakedown of tests)?

hjoliver commented:
It probably would make sense @TomekTrzeciak ... and the longer we leave this the less well tested it will be by the time Cylc 8 is done.  What do you think @dpmatthews ?

oliver-sanders commented:
I got your ping from #3654 

> (looks like a massive shakedown of tests)?

That PR just moves tests into a different location, looks scarier than it is.

#3653 removed the `cylc jobscript` command and replaced the old functional tests with new unit tests so any test changes will need to be ported.

TomekTrzeciak commented:
> I got your ping from #3654
> 
> > (looks like a massive shakedown of tests)?
> 
> That PR just moves tests into a different location, looks scarier than it is.

That might be so, but for someone outside of the core development it's harder to stay on top of these things.

> > I've just noticed a bunch of new conflicts here as a result of #3653
> 
> #3653 removed the `cylc jobscript` command and replaced the old functional tests with new unit tests so any test changes will need to be ported.

I'll wait with it until the review in case more stuff crops up. Also, heads up that I will be out of action in late Aug & Sep, so it would be doubly useful to get this going in case you plan it around that time.

TomekTrzeciak commented:
@oliver-sanders, after quite horrible rebase (I had conflicts from files I haven't even touched in this PR!) this branch should be now up to date with master. All tests pass.

(BTW, it looks like `tests/cylc-get-site-config/06-cylc-conf-path.t` slipped through the cracks in #3654.)

oliver-sanders commented:
Well spotted on the missing test!

TomekTrzeciak commented:
(rebased to resolve conflicts)

TomekTrzeciak commented:
another rebase due conflicts 😱

dpmatthews reviewed (changes requested):
This is looking really good - just a few suggested changes.
Note that I've run lots of tests and reviewed the `job.sh` changes but I'll leave it to other reviewers to check the test changes.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -182,29 +237,45 @@ cylc__job__run_inst_func() {
+    # Propagate real signals to entire process group, if we are a group leader,
+    # otherwise just to the backgrounded user script.
+    if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]] &&
+       [[ ":DEBUG:ERR:EXIT:RETURN:" != *":${signal}:"* ]]; then
```
Discussed offline, changed to `[[ ! "${signal}" =~ '^(ERR|EXIT)$' ]]` to allow any (future) configured signal to be propagated.

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -182,29 +237,45 @@ cylc__job__run_inst_func() {
+    cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" &
     if "${run_err_script}"; then
+        if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]]; then
+            # User script might trap the signal, wait for it before err-script
```
Good point, this way will be more robust and flexible.

TomekTrzeciak commented:
(A lot of tests are failing in Apt-Get Install with Ubuntu image not found, hope someone is looking into this.)

TomekTrzeciak commented on cylc/flow/etc/job.sh:
```diff
@@ -124,10 +122,25 @@ cylc__job__main() {
+            # Trigger ERR trap while preserving the exit code
+            # (NB: bash versions are buggy and neither return statement nor
+            # subshelled exit won't do and we need to spawn another shell).
+            /bin/sh -c "exit $ret_code"
```
done

oliver-sanders commented:
Transitory GH Actions issue, other actions are going through now, re-running.

dpmatthews commented on cylc/flow/etc/job.sh:
```diff
@@ -182,29 +241,42 @@ cylc__job__run_inst_func() {
+    # Propagate real signals to entire process group, if we are a group leader,
+    # otherwise just to the backgrounded user script.
+    if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]] &&
+       [[ ! "${signal}" =~ '^(ERR|EXIT)$' ]]; then
```
The quotes need to be removed from the right-hand side:
https://mywiki.wooledge.org/BashPitfalls#if_.5B.5B_.24foo_.3D.2BAH4_.27some_RE.27_.5D.5D

TomekTrzeciak replied:
@dpmatthews, even though I fixed the quoting, it looks like regex matching is causing bash 3.2 action to fail consistently. I've downloaded the action logs to see what's going on and there is syntax error thrown on this line. Dunno why, because bash 3.2 should support this syntax (it works if I try locally). Should I revert this to use glob matching as before? Or alternatively:

```bash
    if [[ -n "${CYLC_TASK_USER_SCRIPT_PID:-}" ]] &&
       [[ "${signal}" != "ERR" ]] && [[ "${signal}" != "EXIT" ]]; then
```

hjoliver replied:
Ha, that explains it - we were cross-posting :grin:

hjoliver replied:
(Jeez, you must be up late!)

dpmatthews replied:
@TomekTrzeciak I can't explain this either but your alternative suggestion above looks fine to me.

dpmatthews reviewed (approved):
Great - thanks (I've only reviewed the changes to `job.sh`)

hjoliver commented:
(Updated branch and fixed conflicts)

hjoliver reviewed (approved):
Code looks good, very thorough, thanks @TomekTrzeciak  :+1:   Tested as working (several times over the last year :rofl: )

I'll be happy to merge this, just one question for @TomekTrzeciak: Are we expecting the bash-3.2 functional tests to fail?  (Looking back in the GH Actions history, they also failed before I updated and deconflicted the branch).  Or were you planning to make it work on 3.2 as well?  ... @cylc/core - do we really need to support bash 3.2?  (released 2006 :vomiting_face: )

TomekTrzeciak commented:
> I'll be happy to merge this, just one question for @TomekTrzeciak: Are we expecting the bash-3.2 functional tests to fail? 

Fixed, all green now. Please note that there's also associated cylc/cylc-doc#108 that needs reviewing (just updated to deconflict).

hjoliver commented:
Brilliant :tada:  ... ~16 months in the making (apologies @TomekTrzeciak!)
